Move the gRPC support to the main Quarkus repository (#9120)
quarkusio/quarkus: docs/src/main/asciidoc/grpc-getting-started.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////

> = Getting Started with Quarkus gRPC

Reviewer: ```suggestion
= Getting Started with gRPC
```

include::./attributes.adoc[]

> This page explains how to start using Quarkus gRPC.

Reviewer: ```suggestion
This page explains how to start using gRPC in your Quarkus application.
```

While this page describes how to configure it with Maven, it is also possible to use Gradle.

Let's imagine you have a regular Quarkus project, generated from the https://code.quarkus.io[Quarkus project generator].
The default configuration is enough, but you can also select some extensions if you want.

== Configuring your project

Edit the `pom.xml` file to add the quarkus gRPC extension dependency and the JSR 305 (just under `<dependencies>`):

[source, xml]
----
<dependency>
  <groupId>io.quarkus</groupId>
  <artifactId>quarkus-grpc</artifactId>
</dependency>
<dependency>
 <groupId>com.google.code.findbugs</groupId>
 <artifactId>jsr305</artifactId>

>  <scope>provided</scope>

Reviewer: This is annoying because you can't just add the extension.

Author: It's an ongoing discussion with @michalszynkiewicz . I hope that we would be able to remove it from there. Because I agree... should not be required.

Reviewer 2: I agree that requiring users to add a dependency is not the way to go.
Let's drop the ban for jsr305 (at least for now)

Author: Fixed.

                        <id>quarkus-grpc-protoc-plugin</id>
                        <groupId>io.quarkus</groupId>
                        <artifactId>quarkus-grpc-protoc-plugin</artifactId>
                        <version>999-SNAPSHOT</version>
                        <mainClass>io.quarkus.grpc.protoc.plugin.MutinyGrpcGenerator</mainClass>
                    </protocPlugin>
                </protocPlugins>
            </configuration>
            <executions>
                <execution>
                    <id>compile</id>
                    <goals>
                        <goal>compile</goal>
                        <goal>compile-custom</goal>
                    </goals>
                </execution>
                <execution>
                    <id>test-compile</id>
                    <goals>
                        <goal>test-compile</goal>
                        <goal>test-compile-custom</goal>
                    </goals>
                </execution>
            </executions>
        </plugin>

        <!-- ... -->
    </plugins>
</build>
----

The `protobuf-maven-plugin` that generates stub classes from your gRPC service definition (`proto` files).

> The class generation uses a tool named `protoc`, which is OS-specific.

Reviewer: Looks like they should be callouts? At least that's what I suppose from how the sentences are worded?

Author: Fixed.

}

// The request message containing the user's name.
message HelloRequest {
    string name = 1;
}

// The response message containing the greetings
message HelloReply {
    string message = 1;
}
----

This `proto` file defines a simple service interface with a single method (`SayHello`), and the exchanged messages (`HelloRequest` containing the name and `HelloReply` containing the greeting message).

Before coding, we need to generate the classes used to implement and consume gRPC services.
In a terminal, run:

[source, bash]
----
$ mvn compile
----

Once generated, you can look at the `target/generated-sources/protobuf` directory:

[source, txt]
----
target/generated-sources/protobuf
├── grpc-java
│   └── io
│       └── quarkus
│           └── example
│               └── GreeterGrpc.java
└── java
    └── io
        └── quarkus
            └── example
                ├── HelloReply.java
                ├── HelloReplyOrBuilder.java
                ├── HelloRequest.java
                ├── HelloRequestOrBuilder.java
                ├── HelloWorldProto.java
                └── MutinyGreeterGrpc.java
----

These are the classes we are going to use.

> IMPORTANT: Every time you update the `proto` files, you need to re-generate the classes (using `mvn compile`).

Reviewer: How does it work with hot reload?

Author: It does not work yet, @michalszynkiewicz is looking into that. The issue is that your code depends on these classes, so it must be done before the compilation.

== Implementing a gRPC service

Now that we have the generated classes let's implement our _hello_ service.

> With Quarkus gRPC, implementing a service requires to _extend_ the generated service base implementation and exposes it as a `@Singleton` CDI bean.

Reviewer: ```suggestion
With Quarkus, implementing a service requires to _extend_ the generated service base implementation and expose it as a `@Singleton` CDI bean.
```

=== Implementing a service

Create the `src/main/java/org/acme/HelloService.java` file with the following content:

[source, java]
----
package org.acme;

import io.grpc.stub.StreamObserver;
import io.quarkus.example.GreeterGrpc;
import io.quarkus.example.HelloReply;
import io.quarkus.example.HelloRequest;

import javax.inject.Singleton;

@Singleton                                                                                    // <1>
public class HelloService extends GreeterGrpc.GreeterImplBase {                               // <2>

    @Override
    public void sayHello(HelloRequest request, StreamObserver<HelloReply> responseObserver) { // <3>
        String name = request.getName();
        String message = "Hello " + name;
        responseObserver.onNext(HelloReply.newBuilder().setMessage(message).build());         // <4>
        responseObserver.onCompleted();                                                       // <5>
    }
}

----
1. Expose your implementation as bean
2. Extends the `ImplBase` class. This is a generated class.
3. Implement the methods defined in the service definition (here we have a single method)
4. Build and send the response
5. Close the response

> Quarkus also provides an additional model with Mutiny.

Reviewer: maybe add what Mutiny is in 3 words?

Author: Fixed. I also linked the getting-started-with-reactive guide.

[source, java]
----
package org.acme;

import io.quarkus.example.GreeterGrpc;
import io.quarkus.example.HelloRequest;
import io.quarkus.grpc.runtime.annotations.GrpcService;

import javax.inject.Inject;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;

@Path("/hello")
public class ExampleResource {

    @Inject
    @GrpcService("hello")                     // <1>
    GreeterGrpc.GreeterBlockingStub client;   // <2>

    @GET
    @Produces(MediaType.TEXT_PLAIN)
    public String hello() {
        return "hello";
    }

    @GET
    @Path("/{name}")
    public String hello(@PathParam("name") String name) {
        return client.sayHello(HelloRequest.newBuilder().setName(name).build()).getMessage();  // <3>
    }
}
----
1. Inject the service and configure its name. This name is used in the application configuration
2. Use the _blocking_ stub (also a generated class)
3. Invoke the service

We need to configure the application to indicate where is the `hello` service.
In the `src/main/resources/application.properties` file, add the following property:

[source, text]
----

> hello.host=localhost

Reviewer: How is this property mapped? If it's some magic Quarkus config, I would prefer we have some proper `quarkus.` config.

Author: I will have a look tomorrow. When I tried it was not working because it was all made at runtime. But we revisited that, so should work now.

Author: Done.
